community/cypher/cypher-planner-3.4/src/main/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/logical/ReportCostComparisonsAsRows.scala

```diff
@@ -0,0 +1,194 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.compiler.v3_4.planner.logical
+
+import org.neo4j.cypher.internal.compiler.v3_4.phases.LogicalPlanState
+import org.neo4j.cypher.internal.compiler.v3_4.planner.logical.steps.CostComparisonListener
+import org.neo4j.cypher.internal.frontend.v3_4.ast._
+import org.neo4j.cypher.internal.ir.v3_4.PlannerQuery
+import org.neo4j.cypher.internal.planner.v3_4.spi.PlanningAttributes.{Cardinalities, Solveds}
+import org.neo4j.cypher.internal.util.v3_4.attribution.{Id, SequentialIdGen}
+import org.neo4j.cypher.internal.util.v3_4.{Cardinality, Cost, InputPosition}
+import org.neo4j.cypher.internal.v3_4.expressions._
+import org.neo4j.cypher.internal.v3_4.logical.plans._
+
+import scala.collection.{immutable, mutable}
+
+/***
+  * This class can be invoked by writing CYPHER debug=printCosts <query>.
+  *
+  * This class can listen to cost comparisons and report them back as normal rows. This is done by creating a fake plan
+  * that looks something like this:
+  *
+  * UNWIND [{comparison: 0, planId: 1, planText: "plan", planCost: "costs", cost: 12, ...}] as col
+  * RETURN col["comparison] as comparison
+  *
+  * This is the plan that actually gets run, and not the one that the user provided.
+  *
+  * WARNING: This is a debug feature, and as such, not tested as rigorously as other features are.
+  *          Only run this on a system that is not live after taking proper backups!
+  *
+  */
+class ReportCostComparisonsAsRows extends CostComparisonListener {
+
+  private val rows = new mutable.ListBuffer[Row]()
+  private var comparisonCount = 0
+  private val pos = InputPosition(0, 0, 0)
+
+  override def report[X](projector: X => LogicalPlan,
+                         input: Iterable[X],
+                         inputOrdering: Ordering[X],
+                         context: LogicalPlanningContext,
+                         solveds: Solveds,
+                         cardinalities: Cardinalities): Unit = if (input.size > 1) {
+
+    def stringTo(level: Int, plan: LogicalPlan): String = {
+      def indent(level: Int, in: String): String = level match {
+        case 0 => in
+        case _ => "\n" + "  " * level + in
+      }
+
+      val cost = context.cost(plan, context.input, cardinalities)
+      val thisPlan = indent(level, s"${plan.getClass.getSimpleName} costs $cost cardinality ${cardinalities.get(plan.id)}")
+      val l = plan.lhs.map(p => stringTo(level + 1, p)).getOrElse("")
+      val r = plan.rhs.map(p => stringTo(level + 1, p)).getOrElse("")
+      thisPlan + l + r
+    }
+
+    val sortedPlans = input.toIndexedSeq.sorted(inputOrdering).map(projector).reverse
```

Comment:
I know performance is not important here, but sort + reverse is more than I am willing to accept 😀
I believe you can do

```
val sortedPlans = input.toIndexedSeq.sorted(inputOrdering.reverse).map(projector)
```

but since this is the only place where you are using the `inputOrdering` why not just send in the ordering you want instead of its reverse?

Reply:
It's not the only place where it is used, so I'll use your `.reverse` trick instead.